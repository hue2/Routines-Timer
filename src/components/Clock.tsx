import * as React from 'react';
import TimerButton from './Buttons';
//@ts-ignore;
import sound from '../styles/sony_beep_beep_alarm.mp3';

type TimeProps = {
    value?: number,
    hour: number,
    minute: number,
    seconds: number,
    minuteBreak: number,
    secondBreak: number,
    repeats: number,
    handleStartTimer: () => void,
    isBreak?: boolean,
    handlePause: () => void,
    isPaused?: boolean,
    showOptions: boolean,
    onReset: () => void,
    onRestart: () => void,
    remainingRepeats?: number
};

export default class Clock extends React.Component<TimeProps>  {
    formatNumber(hour: boolean, minute: boolean, second: boolean,) {
        let time;
        if (minute) {
            time = this.props.isBreak ? this.props.minuteBreak : this.props.minute;
           
        }
        else if (second) {
            time = this.props.isBreak ? this.props.secondBreak : this.props.seconds;
        }
        else if (hour) {
            time = this.props.hour;
        }

        if (time != undefined && time.toString().length < 1) {
            time = "0";
        }

        return (time != undefined && time.toString().length < 2) ? "0" + time : time;
    }
    render() {
        return(
            <div>
                <div className={this.props.isBreak ? "hidden" : !this.props.showOptions ? "time" : "time padding-200px"} id="time-countdown">
                <b>{this.formatNumber(true, false, false)}:{this.formatNumber(false, true, false)}:{this.formatNumber(false, false, true)}</b>
                </div>
                <div className={!this.props.isBreak ? "hidden" : ""} id ="break-countdown">
                {this.props.isBreak && <audio src={sound} autoPlay />}
                <h3>Break time!</h3>
                <b>
                {this.formatNumber(false, true, false)}:{this.formatNumber(false, false, true)}</b>
                <p className="small-text"># of repeats left: {this.props.remainingRepeats}</p>
                </div>
                {
                 !this.props.showOptions &&
                <TimerButton className={this.props.seconds < 1 
                    && this.props.minute < 1
                    && this.props.hour < 1 ? "disabled-btn " : "primary-btn "} onClick={this.props.handleStartTimer} 
                    disabled={this.props.seconds < 1 
                    && this.props.minute < 1
                    && this.props.hour < 1 ? true : false}
                    id="start-btn">
                    <i className="fa fa-play"></i>&nbsp; Start
                </TimerButton>
                }

                {/* only show options when the timer is started */}
                {this.props.showOptions &&
                <div id="button-groups">
                    <TimerButton className="red-btn"  onClick={this.props.onReset}><i className="fa fa-stop"></i>&nbsp; Reset</TimerButton>
                    {this.props.isPaused &&
                        <TimerButton className="primary-btn" onClick={this.props.handlePause}><i className="fa fa-play"></i>&nbsp; Resume</TimerButton>
                    }
                    {!this.props.isPaused &&
                        <TimerButton className="orange-btn" onClick={this.props.handlePause} id="pause-btn"><i className="fa fa-pause"></i>&nbsp; Pause</TimerButton>
                    }
                    <TimerButton className="gray-btn"  onClick={this.props.onRestart}><i className="fa fa-undo"></i>&nbsp; Restart</TimerButton>            
                </div>
                }
            </div>
        );
    }
}