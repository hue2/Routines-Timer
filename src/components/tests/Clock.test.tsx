import * as React from 'react';
import {shallow} from 'enzyme';
import Clock from '../Clock';
import * as renderer from 'react-test-renderer';

describe('Clock tests', () => {
  let props, wrapper;

  beforeEach(() => {
    props = {
      time:'01:01:01',
      repeats: 0,
      isBreak: false,
      showOptions: false, 
      remainingRepeats: 1,
    }

    wrapper = shallow(
      <Clock
        {...props}
      />
    )
  });

  afterEach(() => {
    wrapper.unmount();
  })

  it('component matches snapshot', () => {
    const tree = renderer.create(<Clock 
      time={''}
      repeats={0}
      isBreak={false}
      showOptions={false}
      remainingRepeats={1}
      />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should reflects value from input', () => {
    expect(wrapper.find('#time-countdown').text()).toBe('01:01:01');
  });

  it('should not show repeat text when no repeats', () => {
    expect(wrapper.find('#repeat-txt').hasClass('hidden')).toEqual(true);
  });

  it('should show repeat text when has repeats', () => {
    const wrapper = shallow(
      <Clock
        {...props}
        repeats={1}
      />
    )
    expect(wrapper.find('#repeat-txt').hasClass('hidden')).toEqual(false);
    wrapper.unmount();
  });

  it('should show clock text when isBreak is false', () => {
    expect(wrapper.find('#time-countdown').hasClass('time')).toEqual(true);
  });

  it('should not show clock text when isBreak is true', () => {
    const wrapper = shallow(
      <Clock
        {...props}
        isBreak={true}
      />
    )
    expect(wrapper.find('#time-countdown').hasClass('time')).toEqual(false);
  });

  it('should show break when isBreak is true', () => {
    const wrapper = shallow(
      <Clock
        {...props}
        isBreak={true}
      />
    )
    expect(wrapper.find('#break-countdown').hasClass('hidden')).toEqual(false);
    wrapper.unmount();
  });
})