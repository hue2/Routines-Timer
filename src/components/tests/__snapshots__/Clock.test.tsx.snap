// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component matches snapshot 1`] = `
<div>
  <div
    className="time padding-200px"
    id="time-countdown"
  >
    <b>
      01:01:01
    </b>
  </div>
  <div
    className="hidden"
    id="break-countdown"
  >
    <h3>
      Break time!
    </h3>
    <b>
      01:01:01
    </b>
    <p
      className="small-text"
    >
      # of repeats left: 
    </p>
  </div>
  <div
    id="button-groups"
  >
    <button
      className="timer-btn reset-btn"
      onClick={[Function]}
    >
      <i
        className="fa fa-times-circle-o"
      />
        Reset
    </button>
    <button
      className="timer-btn pause-btn"
      id="pause-btn"
      onClick={[Function]}
    >
      <i
        className="fa fa-pause-circle-o"
      />
        Pause
    </button>
    <button
      className="timer-btn restart-btn"
      onClick={[Function]}
    >
      <i
        className="fa fa-refresh"
      />
        Restart
    </button>
  </div>
</div>
`;
